### PR TITLE
Add Route Annotation : explicit_defaults

### PR DESCRIPTION
**Pull request by chs2**
Allows generator to add default values to generated URL
in order to get a canonical one

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no 
| Deprecations? | no
| Tests pass?   | yes  
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

If a route has a parameter with default value when it gets generated this parameter is not included

```
news_export:
  path: '/exports/news.{_format}
  requirements:
    _format: xml|json|rss
  defaults:
    _format: xml

$uri = $router -> generate('news_export')
```

`$uri` is `/exports/news` which is identical, from the router point of view, to `/export/news.xml`.

URIs should be unique to avoid duplicate content, which is bad for SEO.

Though it is possible to always add default value to parameters (`$router -> generate('news_export', ['_format' => 'xml', ])`), the added "explicit_defaults" ensures the route will always generated to a canonical form.

### CODE REVIEW

File: src/Symfony/Component/Routing/Route.php [hunk 1]
```diff
@@ -527,6 +530,33 @@ public function setCondition($condition)
...
+     *
+     * This method implements a fluent interface.
+     *
+     * @param string $explicitDefaults The explicitDefaults
+     *
+     * @return $this
+     */
+    public function setExplicitDefaults($explicitDefaults)
+    {
+        $this->explicitDefaults = (bool) $explicitDefaults;
```
Comment (charj) [inline src/Symfony/Component/Routing/Route.php]: You could set this as a boolean in the functions params :)

**Comment by ro0NL** [comment]
what about a new path syntax to do this:

`path: '/exports/news.{>_format?xml}'` (note the `?xml` is already possible, it's about `>` here).

**Comment by nicolas-grekas** [comment]
I like @ro0NL's idea to deal with this at the param syntax level.
I'd suggest using an `!` for that: `/foo.{!bar}` => this would enforce bar to have a default - and would always use it when generating URLs.
WDYT?

**Comment by chs2** [comment]
My original idea was to have a canonical URL generated.
Here the example has only one parameter but there could be more :
```
the_route:
  path: /{section}.{_format}
  requirements:
    section: (index|news|authors)
    _format: (html|xml|json)
  defaults:
    section: index
    _format: html
```

Having the option on the parameter level gives more flexibility but IMO defeats the purpose of the canonical URL.

(BTW, the next step I had in mind was having the Router sending a Redirect when it matches a route on a non-canonical form.)

**Comment by ro0NL** [comment]
we can debate if redirecting non-canonical  to canonical URLs is something that belongs to core :thinking: 

if so, that might also be triggered by e.g. `canonical: true`, which IMHO is a better semantic then `explicit_defaults`.

**Comment by nicolas-grekas** [comment]
The generated URL is already a canonical one to me. I get you'd like it to have the suffix in it instead. It's not about canonical 1/0 to me, but about suffix 1/0, isn't it?

**Comment by nicolas-grekas** [comment]
@chs2 WDYT about these proposals? Would you like to implement it?

**Comment by chs2** [comment]
I'd like to but I can't get my head around the param syntax especially with the XML/XSD.
I could go with "canonical: true" proposed by @ro0NL (instead of my "explicit_defaults").

**Comment by ro0NL** [comment]
@chs2  i dont think we need canonical:true|false, as @nicolas-grekas mentioned each configured route is/should be already a canonical one. So the options are:

- Handle the duplicate content redirect in code / at server level (htaccess) or another router

```yaml
news_export:
  path: '/exports/news.{_format}
  defaults:
    _format: xml
```

- Handle the duplicate content by another route definition

```yaml
news_export_non_canonical:
  path: '/export/news
  controller: Symfony\Bundle\FrameworkBundle\Controller\RedirectController::redirectAction
  defaults:
    route: news_export_xml

news_export_xml:
  path: '/exports/news.xml
```

These are explicit configurations to show we dont need `canonical: true` as a feature. The user should be responsible here.

If we look at option 2, we see our missing step: it cant specify the `{_format}` placeholder to be required (so it would need a route per format).

That's what we should solve by the proposed `{!_format}` syntax, so we can do

```yaml
news_export_non_canonical:
  path: '/export/news
  controller: Symfony\Bundle\FrameworkBundle\Controller\RedirectController::redirectAction
  defaults:
    route: news_export

news_export:
  path: '/exports/news.{!_format}
  defaults:
    _format: xml
```

Due `!` syntax we would always require a format (given or default).

**Comment by ro0NL** [comment]
@chs2  i dont think we need canonical:true|false, as @nicolas-grekas mentioned each configured route is/should be already a canonical one. So the options are:

- Handle the duplicate content redirect in code / at server level (htaccess) or another router

```yaml
news_export:
  path: '/exports/news.{_format}
  defaults:
    _format: xml
```

- Handle the duplicate content by another route definition

```yaml
news_export_non_canonical:
  path: '/export/news
  controller: Symfony\Bundle\FrameworkBundle\Controller\RedirectController::redirectAction
  defaults:
    route: news_export_xml

news_export_xml:
  path: '/exports/news.xml
```

These are explicit configurations to show we dont need `canonical: true` as a feature. The user should be responsible here.

If we look at option 2, we see our missing step: it cant specify the `{_format}` placeholder to be required (so it would need a route per format).

That's what we should solve by the proposed `{!_format}` syntax, so we can do

```yaml
news_export_non_canonical:
  path: '/export/news
  controller: Symfony\Bundle\FrameworkBundle\Controller\RedirectController::redirectAction
  defaults:
    route: news_export

news_export:
  path: '/exports/news.{!_format}
  defaults:
    _format: xml
```

Due `!` syntax we would always require a format (given or default).

**Comment by ro0NL** [comment]
@chs2  i dont think we need canonical:true|false, as @nicolas-grekas mentioned each configured route is/should be already a canonical one. So the options are:

- Handle the duplicate content redirect in code / at server level (htaccess) or another router

```yaml
news_export:
  path: '/exports/news.{_format}
  defaults:
    _format: xml
```

- Handle the duplicate content by another route definition

```yaml
news_export_non_canonical:
  path: '/export/news
  controller: Symfony\Bundle\FrameworkBundle\Controller\RedirectController::redirectAction
  defaults:
    route: news_export_xml

news_export_xml:
  path: '/exports/news.xml
```

These are explicit configurations to show we dont need `canonical: true` as a feature. The user should be responsible here.

If we look at option 2, we see our missing step: it cant specify the `{_format}` placeholder to be required (so it would need a route per format).

That's what we should solve by the proposed `{!_format}` syntax, so we can do

```yaml
news_export_non_canonical:
  path: '/export/news
  controller: Symfony\Bundle\FrameworkBundle\Controller\RedirectController::redirectAction
  defaults:
    route: news_export

news_export:
  path: '/exports/news.{!_format}
  defaults:
    _format: xml
```

Due `!` syntax we would always require a format (given or default).

**Comment by ro0NL** [comment]
@chs2  i dont think we need canonical:true|false, as @nicolas-grekas mentioned each configured route is/should be already a canonical one. So the options are:

- Handle the duplicate content redirect in code / at server level (htaccess) or another router

```yaml
news_export:
  path: '/exports/news.{_format}
  defaults:
    _format: xml
```

- Handle the duplicate content by another route definition

```yaml
news_export_non_canonical:
  path: '/export/news
  controller: Symfony\Bundle\FrameworkBundle\Controller\RedirectController::redirectAction
  defaults:
    route: news_export_xml

news_export_xml:
  path: '/exports/news.xml
```

These are explicit configurations to show we dont need `canonical: true` as a feature. The user should be responsible here.

If we look at option 2, we see our missing step: it cant specify the `{_format}` placeholder to be required (so it would need a route per format).

That's what we should solve by the proposed `{!_format}` syntax, so we can do

```yaml
news_export_non_canonical:
  path: '/export/news
  controller: Symfony\Bundle\FrameworkBundle\Controller\RedirectController::redirectAction
  defaults:
    route: news_export

news_export:
  path: '/exports/news.{!_format}
  defaults:
    _format: xml
```

Due `!` syntax we would always require a format (given or default).

**Pull request closed without merging**